syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,988 @@
+            param_types={"userid": param_types.STRING}
+        ).one()[0]
+        return dt2ts(result) if result else 0
+
+    @with_session
+    def get_collection_timestamps(self, session, user):
+        """Returns the collection timestamps for a user."""
+        userid = user_key(user)
+        res = session.transaction.execute_sql(
+            getq(queries.COLLECTIONS_TIMESTAMPS),
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+        res = self._map_collection_names(res)
+        for collection in res:
+            res[collection] = dt2ts(res[collection])
+        return res
+
+    @with_session
+    def get_collection_counts(self, session, user):
+        """Returns the collection counts."""
+        userid = user_key(user)
+        res = session.transaction.execute_sql(
+            COLLECTIONS_COUNTS,
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+        return self._map_collection_names(res)
+
+    @with_session
+    def get_collection_sizes(self, session, user):
+        """Returns the total size for each collection."""
+        userid = user_key(user)
+        res = session.transaction.execute_sql(
+            COLLECTIONS_SIZES,
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+        # Some db backends return a Decimal() instance for this aggregate.
+        # We want just a plain old integer.
+        rows = ((row[0], int(row[1])) for row in res)
+        return self._map_collection_names(rows)
+
+    @with_session
+    def get_total_size(self, session, user, recalculate=False):
+        """Returns the total size a user's stored data."""
+        userid = user_key(user)
+        size = session.transaction.execute_sql(
+            STORAGE_SIZE,
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        ).one_or_none()
+        return int(size[0] if size else 0)
+
+    @with_session
+    def delete_storage(self, session, user):
+        """Removes all data for the user."""
+        userid = user_key(user)
+
+        # Deleting user_collections will wipe out all BSO's and batches
+        # for the user due to cascading delete
+        session.transaction.execute_update(
+            getq(queries.DELETE_ALL_COLLECTIONS),
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+
+    #
+    # APIs to operate on an individual collection
+    #
+
+    @with_session
+    def get_collection_timestamp(self, session, user, collection):
+        """Returns the last-modified timestamp of a collection."""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        # The last-modified timestamp may be cached on the session.
+        if session.collection_ts:
+            return session.collection_ts
```

Comment:
Does this assume that calling code will only ever access a single collection within the scope of a session?  I wonder if it's worth asserting that explicitly somehow.

Reply:
It does, my reading of the code, specifically with how the lock is done on a per collection basis and access to this variable is only done when working under a single collection lock. This is slightly different to how the cached collection timestamps are handled in non-Spanner, as that use-case benefits from caching all of them since the Python process is the sole accessor of the database and can be aware that the actual values aren't changing out from under it.

I think if we wanted to assert this, we'd cache the collection in question with this timestamp, then verify on access that the collection being access is the same that had its timestamp cached. Note this is a session specific caching of the value though, not on the persistent storage object itself like the cached collection timestamps for non-Spanner.